### PR TITLE
Initialize Cert-manager with ingress certs via ingress-shim workflow

### LINKED ISSUE #35: cert-manager with ingress certs via ingress-shim
We should have a workflow for setting up cert-manager with a ingress controller to test cert issuance via [ingress shim](https://cert-manager.io/docs/usage/ingress/). We could use [Ingress NGINX](https://kubernetes.github.io/ingress-nginx/) for testing this.

The workflow should deploy
- cert-manager
- [ingress NGINX controller](https://kubernetes.github.io/ingress-nginx/deploy/#quick-start)
- a cert-manager issuer, preferably ACME LetsEncrypt
- an `Ingress` resource with [cert-manager annotations to select the deployed issuer](https://cert-manager.io/docs/usage/ingress/#supported-annotations)

[Here](https://github.com/cert-manager/cert-manager/blob/master/make/e2e-setup.mk#L214-L240) is how we set up Ingress NGINX in cert-manager e2e tests. I think we should be able to use the latest ingress NGINX version.
Let me know if you think there will be issues with testing ACME issuer- I can do testing in cloud if needed.

### PR DESCRIPTION
**Pull request by lonelyCZ**
Signed-off-by: lonelyCZ <531187475@qq.com>

Fixes #35

### CODE REVIEW

**Comment by lonelyCZ** [comment]
Hi, @irbekrm I initialize the workflow of deploying ingress, could you help me check whether the workflow is correct? I plan to make it support `ACME` after the initial workflw is correct in next PR.

And I have a question, do we need to deploy an example APP to use the ingress, such as an easy web app?

/assign @irbekrm

File: cm-ingress/common.tfvars [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Path to the kubeconfig file to use for connecting kubernetes cluster. Default is "~/.kube/config"
+kubeconfig_path = "~/.kube/config"
+
+# Cert-manger version to install. Default is "v1.9.1"
+cm_version = "v1.9.1"
+
+# Ingress-nginx version to install. Default is "4.2.3" that is Ingress-nginx Helm version corresponding to Ingress-nginx version "1.3.0".
+# You can check the mapping between the Ingress-nginx Helm version and the Ingress-nginx APP version by executing `helm search repo ingress-nginx/ingress-nginx -l`
```
Comment (irbekrm) [inline cm-ingress/common.tfvars:8]: This is a good tip 👍🏼

File: cm-ingress/common.tfvars [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Path to the kubeconfig file to use for connecting kubernetes cluster. Default is "~/.kube/config"
```
Comment (irbekrm) [inline cm-ingress/common.tfvars:1]: We should also allow to configure `ingress_hostname` at this level as every user will want to provide a custom value for it.

File: cm-ingress/cm-config/main.tf [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+resource "kubernetes_namespace_v1" "ingress-demo" {
+  metadata {
+    name = var.demo_namespace
+  }
+}
+
+resource "kubernetes_manifest" "selfsigned-issuer" {
+  manifest = {
+    "apiVersion" = "cert-manager.io/v1"
+    "kind"       = "ClusterIssuer"
```
Comment (irbekrm) [inline cm-ingress/cm-config/main.tf]: Suggestion: make this an `Issuer`. It only needs to issue one certificate in a namespace that we know.

File: cm-ingress/ingress-config/main.tf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    "spec" = {
+      "rules" = [{
+        "host" = var.ingress_hostname
+        "http" = {
+          "paths" = [{
+            "pathType" = "Prefix"
+            "path"     = "/"
+            "backend" = {
+              "service" = {
+                "name" = "ingress-test-service"
```
Comment (irbekrm) [inline cm-ingress/ingress-config/main.tf]: The name of the service should be parameterized (the user will want to use this Ingress to route traffic to some Service that they have)

File: cm-ingress/ingress-config/main.tf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                "port" = {
+                  "number" = 80
```
Comment (irbekrm) [inline cm-ingress/ingress-config/main.tf]: The port too should be parameterized

File: cm-ingress/common.tfvars [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+# You can check the mapping between the Ingress-nginx Helm version and the Ingress-nginx APP version by executing `helm search repo ingress-nginx/ingress-nginx -l`
+ingress-nginx_version = "4.2.3"
+
+# The arguments will be passed to the cm-manager helm values
+cm_arguments = {
+  installCRDs = "true"
+}
+
+# The arguments will be passed to the ingress-nginx helm values
+ingress_nginx_arguments = {
```
Comment (irbekrm) [inline cm-ingress/common.tfvars:23]: This should also include [load balancer IP](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-helm/#:~:text=controller.service.loadBalancerIP)

**Review by irbekrm — CHANGES_REQUESTED** [review]
Hi @lonelyCZ thank you for working on the Ingress dependency!

This looks good to me overall 👍🏼 

I have added a couple suggestions for parameterization.

I have tested this PR by deploying the default configuration to a Kind cluster and checking that the NGINX Ingress has loaded the correct TLS certificate for the `example.com` hostname using [ingress-nginx kubectl plugin](https://kubernetes.github.io/ingress-nginx/kubectl-plugin/) (`kubectl ingress-nginx certs --host example.com --namespace ingress-nginx`). I saw that the correct cert was loaded 👍🏼 
I will test end to end with an external IP once we have an ACME issuer.

**Comment by irbekrm** [comment]
> And I have a question, do we need to deploy an example APP to use the ingress, such as an easy web app?

I think it is OK without as long as we parameterize the [backend service name](https://github.com/cert-manager/testing-addons/pull/37#discussion_r959158940) and port. The user can then deploy whichever app with a Kubernetes Service they want.
Maybe we can add a link to [example kuard deployment](https://cert-manager.io/docs/tutorials/acme/nginx-ingress/#step-4---deploy-an-example-service) in the Readme for cm-ingress module.

**Comment by lonelyCZ** [comment]
PTAL. 

/assign @irbekrm

**Review by irbekrm — APPROVED** [review]
Thank you for making the changes @lonelyCZ , this all now makes sense 👍🏼 

Thanks for adding the Readme too, it's very good.

I guess the next step will be to add the ACME issuer.

/lgtm

**Pull request merged**